### ISSUE TITLE
Softdeleteable and Timestampable together

### ISSUE THREAD

**Issue by pauljura** [doctrine-extensions/DoctrineExtensions#2578]
Hi,

I'm using Softdeleteable and Timestampable together on an entity that has a `dateUpdated` property and a `dateDeleted` property. The problem is that when I soft-delete this, the `dateUpdated` doesn't update, which I expect it to since since Softdeleteable is effectively updating the entity.

I'm using this extension in a Symfony project with `stof/doctrine-extensions-bundle` and looking at similar issues the recommendation seems to be to change the priority of listeners, but I've tried and nothing seems to work.

I've resorted to writing a custom listener that listens to the `preSoftDelete` event in order to set `dateUpdated` but that just feels wrong.

Any advice? Thanks.

**Comment by nayodahl** [comment]
Facing the same issue here, any progress @pauljura ?

**Comment by pauljura** [comment]
Here's my event listener that you can add to your project. It's probably not the best or most efficient code but it works for me.

```
class UpdateOnSoftDelete
{
    public function preSoftDelete(LifecycleEventArgs $eventArgs): void
    {
        $objectManager = $eventArgs->getObjectManager();
        $unitOfWork = $objectManager->getUnitOfWork();

        foreach ($unitOfWork->getScheduledEntityDeletions() as $entity) {
            $className = get_class($entity);
            $propertiesToUpdate = [];
            $reflectionClass = new \ReflectionClass($className);
            $properties = $reflectionClass->getProperties();
            // Loop over all properties of the entity
            foreach ($properties as $property) {
                $attributes = $property->getAttributes();
                // For each property, loop over all the attributes of that property
                foreach ($attributes as $attribute) {
                    // Skip attributes that are not the Timestampable attribute
                    if ($attribute->getName() != Timestampable::class) {
                        continue;
                    }
                    /** @var Timestampable $instance */
                    $instance = $attribute->newInstance();
                    if ($instance->on === 'update') {
                        // We have found a property that has the attribute #[Timestampable(on: 'update')]
                        $oldValue = $property->getValue($entity);
                        $newValue = new \DateTime();
                        $property->setValue($entity, $newValue);
                        $unitOfWork->scheduleExtraUpdate($entity, [
                            $property->getName() => [$oldValue, $newValue],
                        ]);
                    }
                }
            }
        }
    }
}
```

Then add to your services.yaml file:
```
services:
    EventListener\UpdateOnSoftDelete:
        tags:
            -
                name: 'doctrine.event_listener'
                event: 'preSoftDelete'
                priority: 300
                connection: 'default'
```

**Comment by nayodahl** [comment]
thanks @pauljura for your help

**Comment by pauljura** [comment]
Yes, you're right, I thought about memoizing the properties so it wouldn't need to re-evaluate on every entity, but then changed my mind to keep it simple :)

Thanks for the tip about #[AsDoctrineListener] !

**Comment by pauljura** [comment]
Bump. This is still an issue.

**Comment by mbabker** [comment]
Don't stale...

> I'm using this extension in a Symfony project with stof/doctrine-extensions-bundle and looking at similar issues the recommendation seems to be to change the priority of listeners, but I've tried and nothing seems to work.

I don't use soft-delete so I'm not sure how to look into this, but, if anyone's got any suggestions for the documentation on putting listeners in the right order, that would definitely be a helpful starting point.

Also keep in mind that Doctrine's event manager natively doesn't have a `priority` option, in the context of Symfony the `DoctrineBundle` sorts the configured listeners when creating the event manager service.  So we might need something explaining this setup in the context of standalone Doctrine usage and within Symfony.

The bundle registers extensions in alphabetical order, and looking over the two listeners here, I don't immediately see any issue with `SoftDeleteableListener::onFlush()` running before `TimestampableListener::onFlush()` that should keep the timestampable listener from updating that field when soft-delete runs.  But, I find [this block](https://github.com/doctrine-extensions/DoctrineExtensions/blob/e85560ed96f977b8c29428a99222cb2ef2f0e80d/src/SoftDeleteable/SoftDeleteableListener.php#L110-L116) in the soft-delete listener to be a bit interesting as the way it pushes updates up through the unit of work for each object manager is different, so the way it's updating the unit of work for the ORM may be causing the extra changes here not to be visible when the timestampable listener runs since scheduled extra updates aren't pushed into the object changeset.
